src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -2294,6 +2305,82 @@ fn suggest_disambiguator(
     }
 }
 
+/// Report a link to an item that will not have documentation generated.
+fn missing_docs_for_link(
+    cx: &DocContext<'_>,
+    item: &Item,
+    destination_id: DefId,
+    why: HrefError,
+    path_str: &str,
+    diag_info: &DiagnosticInfo<'_>,
+) {
+    // FIXME: this is a bug, rustdoc should load all items into the cache before doing this check.
+    //   Otherwise, there will be false negatives where rustdoc doesn't warn but should.
+    if why == HrefError::NotInExternalCache {
+        return;
+    }
+
+    let item_name = match item.name {
+        Some(name) => name.to_string(),
+        None => "<unknown>".into(),
+    };
+    let msg = format!(
+        "documentation for `{}` links to item `{}` which will not have documentation generated",
+        item_name, path_str
+    );
+
+    report_diagnostic(cx.tcx, PRIVATE_INTRA_DOC_LINKS, &msg, diag_info, |diag, sp| {
+        use crate::clean::types::{AttributesExt, NestedAttributesExt};
+
+        if let Some(sp) = sp {
+            diag.span_label(sp, "this item is will not be documented");
+        }
+
+        if why == HrefError::DocumentationNotBuilt {
+            diag.note(&format!(
+                "`{}` is in the crate `{}`, which will not be documented",
+                path_str,
+                cx.tcx.crate_name(destination_id.krate)
+            ));
+            return;
+        }
+
+        // NOTE: theoretically it shouldn't be possible to resolve private items,
+        // but `resolve_primitive_associated_item` is buggy and allows it.
```

Comment:
This comment is wrong - the actual issue is that primitive associated items are always reachable, but `href` doesn't take that into account